llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,285 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
+    strcpy(newSymName, rodataSymbol.pSymName);
+    strcat(newSymName, CachedRodataSymbolSuffix);
+    symbol.secIdx = m_sections.size() - 1;
+    symbol.secName = secName;
+    symbol.size = rodataSymbol.size;
+    symbol.pSymName = newSymName;
+    symbol.value = rodataSymbol.value;
+    symbol.info.all = rodataSymbol.info.all;
+    symbol.nameOffset = InvalidValue;
+    m_symbols.push_back(symbol);
+  }
+}
+
+// =====================================================================================================================
+// Get the reloc symbol index based on the input reloc symbol
+//
+// @param inputRelocSymbol : The input reloc symbol
+template <class Elf> uint32_t ElfWriter<Elf>::getRelocSymbolIndex(ElfSymbol inputRelocSymbol) {
+
+  char *newSymName = new char[strlen(inputRelocSymbol.pSymName) + strlen(CachedRodataSymbolSuffix)];
+  strcpy(newSymName, inputRelocSymbol.pSymName);
+  strcat(newSymName, CachedRodataSymbolSuffix);
+
+  uint32_t index = 0;
+  for (auto symbol : m_symbols) {
+    if (symbol.secIdx == InvalidValue)
+      continue;
+    if (strcmp(newSymName, symbol.pSymName) == 0)
+      break;
+    index++;
+  }
+  return index;
+}
+
+// =====================================================================================================================
+// Merge base reloc section and input reloc section into merged section
+//
+// @param reader : The ElfReader to the input fragment elf bin
+// @param pSection1 : The first reloc section buffer to merge
+// @param section1RelocsCount : Relocs count of the first section
+// @param pSection2 : The second reloc section buffer to merge
+// @param section2RelocsOffset : Relocs count offset of the second section
+// @param diffOfPsOffset : The delta of ps Isa offset of the two sections
+// @param [out] pNewSection : Merged section
+template <class Elf>
+void ElfWriter<Elf>::mergeRelocSection(const ElfReader<Elf> &reader, const SectionBuffer *pSection1,
+                                       size_t section1RelocsCount, const SectionBuffer *pSection2,
+                                       size_t section2RelocsOffset, int64_t diffOfPsOffset,
+                                       SectionBuffer *pNewSection) {
+
+  *pNewSection = *pSection1;
+
+  size_t section1Size = section1RelocsCount * pSection1->secHead.sh_entsize;
+
+  size_t newSectionSize = section1Size;
+  size_t relocsCount = section1RelocsCount;
+
+  if (pSection2) {
+    size_t section2Offset = section2RelocsOffset * pSection2->secHead.sh_entsize;
+    newSectionSize += (pSection2->secHead.sh_size - section2Offset);
+    relocsCount += (getRelocsCount(pSection2) - section2RelocsOffset);
+  } else if (section1RelocsCount == 0) {
+    // The second section is null and the relocs count of the first section is zero,
+    // no data need to be merged, return a null section.
+    pNewSection->data = nullptr;
+    pNewSection->secHead.sh_size = 0;
+    return;
+  }
+
+  auto mergedData = new uint8_t[newSectionSize];
```

Comment:
Where is this memory freed?

Reply:
This memory will be freed when calling `~ElfWriter()`.

Reply:
Can you mention that in comments next to each `new`?

Reply:
sure.